src/UserSpaceInstrumentation/InjectLibraryInTracee.cpp

```diff
@@ -0,0 +1,298 @@
+  FAIL_IF(result_return_value.has_error(), "Unable to read registers after function called :\"%s\"",
+          result_return_value.error().message());
+  return return_value_registers.GetGeneralPurposeRegisters()->x86_64.rax;
+}
+
+// Execute the code at `address_code`. The code segment has to end with an `int3`.
+void ExecuteOrDie(pid_t pid, RegisterState& original_registers, uint64_t address_code) {
+  RegisterState registers_set_rip = original_registers;
+  registers_set_rip.GetGeneralPurposeRegisters()->x86_64.rip = address_code;
+  auto result_restore_registers = registers_set_rip.RestoreRegisters();
+  if (result_restore_registers.has_error()) {
+    FATAL("Unable to continue tracee with PTRACE_CONT.");
+  }
+  if (ptrace(PTRACE_CONT, pid, 0, 0) != 0) {
+    FATAL("Unable to continue tracee with PTRACE_CONT.");
+  }
+  int status = 0;
+  pid_t waited = waitpid(pid, &status, 0);
+  if (waited != pid || !WIFSTOPPED(status) || WSTOPSIG(status) != SIGTRAP) {
+    FATAL("Failed to wait for sigtrap after PTRACE_CONT.");
+  }
+}
+
+}  // namespace
+
+[[nodiscard]] ErrorMessageOr<void*> DlopenInTracee(pid_t pid, std::filesystem::path path,
+                                                   uint32_t flag) {
+  // Figure out address of dlopen in libc.
+  OUTCOME_TRY(address_dlopen, FindFunctionAddress(pid, "dlopen", "libc-"));
+
+  // Backup registers.
+  RegisterState original_registers;
+  OUTCOME_TRY(original_registers.BackupRegisters(pid));
+
+  // Allocate small memory area in the tracee. This is used for the code and the path name.
+  const uint64_t path_length = path.string().length() + 1;  // Include terminating zero.
+  const uint64_t memory_size = kScratchPadSize + path_length;
+  OUTCOME_TRY(address_code, AllocateInTracee(pid, memory_size));
+
+  // Write the name of the .so into memory at address_code with offset of kScratchPadSize.
+  std::vector<uint8_t> path_as_vector(path_length);
+  memcpy(path_as_vector.data(), path.c_str(), path_length);
+  const uint64_t address_so_path = address_code + kScratchPadSize;
+  auto result_write_path = WriteTraceesMemory(pid, address_so_path, path_as_vector);
+  if (result_write_path.has_error()) {
+    FreeMemoryOrDie(pid, address_code, memory_size);
+    return result_write_path.error();
+  }
+  // We want to do the following in the tracee:
+  // return_value = dlopen(path, RTLD_LAZY);
+  // The calling convention is to put the parameters in registers rdi and rsi.
+  // So the address of the file path goes to rdi. The flag argument goes into rsi (RTLD_LAZY is
+  // equal to 1). Then we load the address of dlopen into rax and do the call. Assembly in Intel
+  // syntax (destination first), machine code on the right:
+
+  // movabs rax, address_so_path      48 b8 address_so_path
+  // mov rbx, 1                       48 c7 c3 01 00 00 00
+  // mov rdi, rax                     48 8b f8
+  // mov rsi, rbx                     48 8b f3
```

Comment:
What's the reason for loading into `rax` and `rbx` just to immediately copy their content into `rdi` and `rsi` (and also immediately overwriting `rax`)? Can't you load directly into `rdi` and `rsi`?